### PR TITLE
Add Manacher’s Algorithm for Longest Palindromic Substring

### LINKED ISSUE #5453: [FEATURE REQUEST] <title>Add a new Algorithm Manacher’s Algorithm
### What would you like to Propose?

Here's an algorithm that is not listed in the structure you provided: "Manacher's Algorithm", which is used to find the longest palindromic substring in linear time.

Algorithm: Manacher’s Algorithm
Purpose:
Given a string s, the algorithm finds the longest palindromic substring in linear time (O(n)).

### Issue details

Preprocess the string:

Insert special characters (like #) between each character of the original string and add ^ at the beginning and $ at the end to avoid boundary conditions. This turns any even-length palindrome into an odd-length one.
Example: "abba" becomes "^#a#b#b#a#$".
Initialize Variables:

P[]: An array where P[i] stores the radius of the palindrome centered at position i.
C: The center of the current palindrome.
R: The right boundary of the current palindrome.
Iterate Over the String:

For each character in the transformed string:
Mirror the palindrome from the left side if the current index is within the bounds of the current palindrome.
Try expanding the palindrome centered at the current character.
Update the center C and the right boundary R if the palindrome expands beyond the current boundary.
Return the Result:

The longest palindrome found will be the one with the maximum value in P[]. Use this to find the center and reconstruct the longest palindromic substring from the original string.

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by donutz03**
- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] All new algorithms have a URL in their comments that points to Wikipedia or other similar explanations.
- [x] All new code is formatted with `clang-format -i --style=file path/to/your/file.java`

This PR adds Manacher's Algorithm to find the longest palindromic substring in linear time.

Closes #5453

### CODE REVIEW

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+     * Finds the longest palindromic substring using Manacher's Algorithm
+     *
+     * @param s The input string
+     * @return The longest palindromic substring in {@code s}
+     */
+    public static String longestPalindrome(String s) {
+        // Preprocess the string to avoid even-length palindrome issues
+        String processedString = preprocess(s);
+        int n = processedString.length();
+        int[] p = new int[n]; // Array to store the radius of palindromes
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/strings/Manacher.java:32]: What do you think about renaming `p` into something meaningful, and remove this comment?

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+            // Track the maximum length and center index of the longest palindrome found so far
+            if (p[i] > maxLen) {
+                maxLen = p[i];
+                centerIndex = i;
+            }
+        }
+
+        // Extract the longest palindrome from the original string
+        int start = (centerIndex - maxLen) / 2; // Get the starting index in the original string
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/strings/Manacher.java:69]: Better to make it `final int start = (centerIndex - maxLen) / 2;` without a comment

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        int n = processedString.length();
+        int[] p = new int[n]; // Array to store the radius of palindromes
+
+        // Separate variable declarations into individual statements
+        int center = 0;
+        int rightBoundary = 0;
+        int maxLen = 0;
+        int centerIndex = 0;
+
+        // Iterate over the preprocessed string to calculate the palindrome radii
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/strings/Manacher.java:40]: What do you think about removing this comment? It seems to describe actions that are already clear from the code.

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+            }
+
+            // Track the maximum length and center index of the longest palindrome found so far
+            if (p[i] > maxLen) {
+                maxLen = p[i];
+                centerIndex = i;
+            }
+        }
+
+        // Extract the longest palindrome from the original string
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/strings/Manacher.java:68]: The same, what do you think about removing this comment? It seems to describe actions that are already clear from the code also.

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Finds the longest palindromic substring using Manacher's Algorithm
+     *
+     * @param s The input string
+     * @return The longest palindromic substring in {@code s}
+     */
+    public static String longestPalindrome(String s) {
+        // Preprocess the string to avoid even-length palindrome issues
+        String processedString = preprocess(s);
+        int n = processedString.length();
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/strings/Manacher.java:31]: What do you think about using `processedString.length()` directly? It's only used in two places, so a separate variable might not be necessary.

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+    /**
+     * Finds the longest palindromic substring using Manacher's Algorithm
+     *
+     * @param s The input string
+     * @return The longest palindromic substring in {@code s}
+     */
+    public static String longestPalindrome(String s) {
+        // Preprocess the string to avoid even-length palindrome issues
+        String processedString = preprocess(s);
```
Comment (alxkm) [inline src/main/java/com/thealgorithms/strings/Manacher.java:30]: Better to use final here. `final String processedString = preprocess(s);`

**Review by alxkm — CHANGES_REQUESTED** [review]

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 7]
```diff
@@ -0,0 +1,91 @@
+package com.thealgorithms.strings;
+
+/**
+ * Wikipedia: https://en.wikipedia.org/wiki/Longest_palindromic_substring#Manacher's_algorithm
+ */
+public final class Manacher {
+
+    // Private constructor to prevent instantiation
```
[the remark this answers is not in the document]
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:8]: removed description

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    private Manacher() {
+    }
+
+    /**
+     * Test code for Manacher's Algorithm
+     */
+    public static void main(String[] args) {
```
[the remark this answers is not in the document]
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:15]: Removed main method

File: src/test/java/com/thealgorithms/strings/ManacherTest.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
+package com.thealgorithms.strings;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+public class ManacherTest {
```
[the remark this answers is not in the document]
Comment (donutz03) [reply src/test/java/com/thealgorithms/strings/ManacherTest.java:7]: Added Parameterized tests and formatted.

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+     * @param s The input string
+     * @return The longest palindromic substring in {@code s}
+     */
+    public static String longestPalindrome(String s) {
+        // Preprocess the string to avoid even-length palindrome issues
+        String processedString = preprocess(s);
+        int n = processedString.length();
+        int[] p = new int[n]; // Array to store the radius of palindromes
+
+        // Separate variable declarations into individual statements
```
[the remark this answers is not in the document]
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:34]: removed comment

File: src/main/java/com/thealgorithms/strings/Manacher.java (hunk 2 above)
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:69]: removed comment, made variable start final

File: src/main/java/com/thealgorithms/strings/Manacher.java (hunk 1 above)
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:32]: Renamed from p to palindromeLengths and removed comment

File: src/main/java/com/thealgorithms/strings/Manacher.java (hunk 5 above)
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:31]: Removed variable n, replaced with preprocessedString.length()

File: src/main/java/com/thealgorithms/strings/Manacher.java (hunk 6 above)
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:30]: updated to use final

File: src/main/java/com/thealgorithms/strings/Manacher.java (hunk 4 above)
Comment (donutz03) [reply src/main/java/com/thealgorithms/strings/Manacher.java:68]: Removed the comment

**Review by donutz03 — COMMENTED** [review]
Resolved all comments. Changed files to support review.

File: src/main/java/com/thealgorithms/strings/Manacher.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+    private Manacher() {
+    }
+
+    /**
+     * Finds the longest palindromic substring using Manacher's Algorithm
+     *
+     * @param s The input string
+     * @return The longest palindromic substring in {@code s}
+     */
+    public static String longestPalindrome(String s) {
```
Comment (donutz03) [inline src/main/java/com/thealgorithms/strings/Manacher.java:17]: Please review when available.

File: src/test/java/com/thealgorithms/strings/ManacherTest.java [hunk 12]
```diff
@@ -0,0 +1,53 @@
+package com.thealgorithms.strings;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import java.util.stream.Stream;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+public class ManacherTest {
```
Comment (donutz03) [inline src/test/java/com/thealgorithms/strings/ManacherTest.java:10]: Please review this class when available.

**Review by alxkm — APPROVED** [review]

**Comment by alxkm** [comment]
Thank you for your contribution

**Pull request merged**